cartography/intel/lastpass/users.py

```diff
@@ -0,0 +1,70 @@
+import logging
+from typing import Dict
+from typing import List
+
+import neo4j
+from dateutil import parser as dt_parse
+from requests import Session
+
+from cartography.client.core.tx import load_graph_data
+from cartography.graph.querybuilder import build_ingestion_query
+from cartography.intel.lastpass.schema import LastpassUserSchema
+from cartography.util import timeit
+
+logger = logging.getLogger(__name__)
+
+
+@timeit
+def sync(
+    neo4j_session: neo4j.Session,
+    update_tag: int,
+    lastpass_cid: str,
+    lastpass_provhash: str,
+) -> None:
+    users = get(lastpass_cid, lastpass_provhash)
+    formated_users = transform(users)
+    load(neo4j_session, formated_users, update_tag)
+
+
+@timeit
+def get(lastpass_cid: str, lastpass_provhash: str) -> dict:
+    payload = {
+        'cid': lastpass_cid,
+        'provhash': lastpass_provhash,
+        'cmd': 'getuserdata',
+        'data': None,
+    }
+    session = Session()
+    req = session.post('https://lastpass.com/enterpriseapi.php', data=payload, timeout=20)
+    req.raise_for_status()
+    return req.json()
+
+
+@timeit
+def transform(api_result: dict) -> List[Dict]:
+    result: List[dict] = []
+    for uid, user in api_result['Users'].items():
+        n_user = user.copy()
+        n_user['id'] = int(uid)
+        for k in ('created', 'last_pw_change', 'last_login'):
+            if n_user[k] == '':
```

Comment:
But if the string is empty in `n_user[k]`, would it also be empty in `user[k]`? Does the `parse(user[k]`) raise an exception?

Reply:
Sure it raises an Exception

Reply:
It seems like lines 50 and 51 are actually unnecessary because the exception to be raised is unhandled.

Reply:
On the contrary is because the error is hunandled that we have to do this check

A user who never logged in will have an empty last_login, resulting on a hunandled exception that will crash the tool.

Reply:
From your testing, the Lastpass API returns empty strings?

If this is the case, I think it is ok to do 

```python
n_user[k] = int(dt_parse.parse(n_user[k]).timestamp() * 1000) if n_user[k] else None
```
because the Neo4j driver will automatically leave that field as unset on the node, and because the empty string is Falsy in Python.

So overall, I think we can we write this function as 

```python
@timeit
def transform(api_result: Dict[str, Any]) -> List[Dict[str, Any]]:
    result: List[Dict[str, Any]] = []
    for uid, user in api_result['Users'].items():
        n_user = user.copy()
        n_user['id'] = int(uid)
        for k in ('created', 'last_pw_change', 'last_login'):
            n_user[k] = int(dt_parse.parse(n_user[k]).timestamp() * 1000) if n_user[k] else None
        result.append(n_user)
    return result
```

Reply:
@ramonpetgrave64 my bad I just checked my code again and there is a missing 'else' (or continu in the if condition) ... I known understand your point ;)

@achantavy your solution is more elegant thant mine